test/test_modules.py

```diff
@@ -156,6 +156,94 @@ def test_pickle(self, device, dtype, module_info):
                     output_from_copy = m_copy(*args, **kwargs)
                     self.assertEqual(output, output_from_copy)
 
+    @modules(module_db, allow_devices={"cuda"})
+    def test_cpu_gpu_results_are_equal(self, device, dtype, module_info):
+        # Test cpu and gpu results are the same
+        module_cls = module_info.module_cls
+        module_inputs_cpu = module_info.module_inputs_func(module_info, device="cpu", dtype=dtype,
+                                                           requires_grad=True)
+
+        def _make_leafs(item):
+            if isinstance(item, dict):
+                for i in item.values():
+                    _make_leafs(i)
+            elif isinstance(item, tuple):
+                for i in item:
+                    _make_leafs(i)
+            else:
+                if not isinstance(item, torch.Tensor) or item.is_leaf:
+                    return
+                old_requires_grad = item.requires_grad
```

Comment:
Hm, this is interesting. Is this needed because of NLLLoss's ModuleInputs? Looks like the input there is created with `requires_grad=True` then `log_softmax()` is called, creating a graph which breaks the test.

I wonder if we want this somewhere more general to enforce that all tensors with `requires_grad=True` coming from `module_inputs_func`s are guaranteed to have an empty graph. Wouldn't this be needed for the gradcheck tests as well?

Also if we do make it more general and called automagically, it probably should handle lists as well.

Reply:
> Wouldn't this be needed for the gradcheck tests as well?

`gradcheck` will check the input and call `retain_grad`:

https://github.com/pytorch/pytorch/blob/c12df2dc2358e58c1dc66221e18635fe29f19e06/torch/autograd/gradcheck.py#L663

I updated this function to do the same.